### PR TITLE
Update GObject annotations for libvips/conversion directory

### PR DESCRIPTION
**Pull request by astavale**
Further work on #741 

Two queries came up when looking through `libvips/conversion`:

 1. Should `vips_bandjoin_const()` in `bandjoin.c` be `**in` instead of `*in` ? The docs state it is an array.

 2. The annotation for `vips_bandrank( VipsImage **in, VipsImage **out, int n, ... )` in `bandrank.c` is ` @in: (array length=n): array of input images`. Should this include `(tansfer none)`, similar to `vips_arrayjoin()` ?

### CODE REVIEW

**Comment by jcupitt** [comment]
Hi again, 

1. Nice find, this is a copy-paste error in the docs, it should just be a single image.

2. Sure, add a transfer-none, though I think that's the default.

**Comment by jcupitt** [comment]
It's great you are adding this extra markup, I think it will improve the docs, but I'm worried you won't get what you want. I don't think adding this markup will give you a working Vala binding :( 

As I said, I think you will need to use the layer below with the (new / set / build / get) machinery.

**Comment by jcupitt** [comment]
This set of changes look fine.

File: libvips/conversion/bandjoin.c [hunk 1]
```diff
@@ -428,9 +428,9 @@ vips_bandjoin_constv( VipsImage *in, VipsImage **out,
 }
 
 /**
- * vips_bandjoin_const:
- * @in: (array length=n) (transfer none): array of input images
- * @out: output image
+ * vips_bandjoin_const: (method)
+ * @in: array of input images
```
Comment (jcupitt) [inline libvips/conversion/bandjoin.c]: Could you fix the comment line for @in as well? Something like "input image"

Comment (astavale) [reply libvips/conversion/bandjoin.c]: My apologies, I had done that, but I had re-based against master that caused a problem and had to make the changes again.

**Comment by astavale** [comment]
The Vala binding is working, but could be more elegant. Vala has a tool, `vapigen`, that uses the GIR file to generate a VAPI file. The VAPI provides translation details to the Vala compiler so the right C code is generated for the bound library. If you've tried this yourself then the base class ObjectClass needed to be adjusted to inherit from GLib.ObjectClass. Also ArgumentTable needed to inherit from GLib.HashTable. I'll publish what I've done so far in the coming weeks.

There are a couple of advantages to using the GIR:

 * The VAPI is generated automatically with the `vapigen` tool, otherwise a VAPI needs to be written manually
 * The documentation for the VAPI is generated automatically. The Vala community have a process where GIRs are held in a public repository, https://github.com/nemequ/vala-girs/tree/master/gir-1.0, and then the documentation is re-generated periodically and appears on the Vala website https://valadoc.org/ I was hoping libvips would fit in to that process

I understand you wish to release in September. So I'm updating the annotations to hopefully fit in with that. This gives a workable binding for people to try out. Given time constraints I'm looking to update libvips/iofuncs and libvips/resample. That should then cover the most common use cases. I may consider a few more directories if I get the time.

Long term, I'm thinking there should probably be a GObject Introspection binding. While Vala can handle varargs, PyGObject doesn't. See - https://wiki.gnome.org/Projects/PyGObject/IntrospectionPorting#Non-introspectable_functions.2BAC8-methods . PyGObject does however allow properties to be set at construct time, which is like keyword args. See - https://wiki.gnome.org/Projects/PyGObject/IntrospectionPorting#Constructors Vala allows for an Object( ... ) keyword, but the only examples I'm finding are in the constructor itself. So I need to investigate that further. I'm mentioning Python here because I understand the typelib file is only generated from a GIR.

**Comment by jcupitt** [comment]
Oh OK, perhaps it's not an issue. I was concerned that you would be unable to use the optional arguments from Vala, but if it can call varargs functions, perhaps it's going to work. 

There's a Python binding based on pygobject here, you probably saw:

https://github.com/jcupitt/libvips/tree/master/python

But it's rather unsatisfactory for various reasons, so I'm planning to deprecate it and push people towards a plain FFI binding instead.

**Pull request merged**